### PR TITLE
Parallel block body download

### PR DESCRIPTION
**Pull request by arkpar**
More aggressive parallel downloading of block bodies.

### CODE REVIEW

**Comment by gavofyork** [comment]
not sure i fully understand the logic, but does seem to be doing roughly the right thing.

**Comment by arkpar** [comment]
Currently we only download bodies for the headers in the first subchain. This change allows idle peers to download all missing bodies, while still prioritizing those in the front of the queue.

**Comment by arkpar** [comment]
I've been testing this branch on zero.parity.io for 2 days, no issues so far.

File: sync/src/blocks.rs [hunk 1]
```diff
@@ -120,13 +120,22 @@ impl BlockCollection {
...
 			}
 		}
-		self.downloading_bodies.extend(needed_bodies.iter());
+		for h in self.header_ids.values() {
+			if needed_bodies.len() >= count {
+				break;
+			}
+			if !self.downloading_bodies.contains(h) {
+				needed_bodies.push(h.clone());
+				self.downloading_bodies.insert(h.clone());
```
Comment (rphmeier) [inline sync/src/blocks.rs]: entry API could save a HashMap lookup here

Comment (arkpar) [inline sync/src/blocks.rs]: HashSet does not have one

**Pull request merged**